### PR TITLE
Fix/40149: add data-* attributes autocomplete

### PR DESCRIPTION
**Pull request by pfongkye**
This issue is referenced here: https://github.com/Microsoft/vscode/issues/40149

### CODE REVIEW

File: src/services/htmlCompletion.ts [hunk 1]
```diff
@@ -175,9 +175,41 @@ export function doComplete(document: TextDocument, position: Position, htmlDocum
...
+		collectDataAttributesSuggestions(range);
 		return result;
 	}
 
+	function collectDataAttributesSuggestions(range: Range) {
+
+		const dataAttr = 'data-';
+		let dataAttributes: string[] = [];
+
+		function addNodeDataAttributes(node: Node) {
```
Comment (pfongkye) [inline src/services/htmlCompletion.ts]: I could try a non-recursive approach

File: src/services/htmlCompletion.ts [hunk 2]
```diff
@@ -175,9 +175,41 @@ export function doComplete(document: TextDocument, position: Position, htmlDocum
...
+			node.attributeNames.filter(attr => attr.substr(0, dataAttr.length) === dataAttr).forEach(attr => {
```
Comment (aeschli) [inline src/services/htmlCompletion.ts]: I'd suggest doing a single forEach loop (no `filter`)
use `startsWith` from `../utils/strings` for the check.

File: src/services/htmlCompletion.ts [hunk 3]
```diff
@@ -175,9 +175,41 @@ export function doComplete(document: TextDocument, position: Position, htmlDocum
...
+				if (dataAttributes.indexOf(attr) === -1) {
```
Comment (aeschli) [inline src/services/htmlCompletion.ts]: Use a object or Map to check for extistece
```ts
if (!dataAttributes['attr']) {
    dataAttributes['attr'] = true;
    ...
```

File: src/services/htmlCompletion.ts (hunk 2 above)
Comment (pfongkye) [reply src/services/htmlCompletion.ts]: thanks for the review 👍

File: src/services/htmlCompletion.ts (hunk 3 above)
Comment (pfongkye) [reply src/services/htmlCompletion.ts]: ok I'll make the modification ASAP

**Comment by pfongkye** [comment]
@aeschli requested modifications applied.

**Comment by aeschli** [comment]
@pfongkye Thanks, great PR!

**Pull request merged**